File: backend/goat_catalog/src/goat_catalog/tokens/otps/application/use_cases/confirm_email_use_case.py
"""Caso de uso para confirmar un email (verificar estado)."""

from ..dto.confirm_email_response import ConfirmEmailResponse
from ...domain.repositories.confirmed_email_repository import ConfirmedEmailRepository
from ...domain.value_objects.email import Email

class ConfirmEmailUseCase:
    """Caso de uso para verificar si un email está confirmado."""

    def __init__(self, confirmed_email_repository: ConfirmedEmailRepository) -> None:
        """Inicializa el caso de uso.

        Args:
            confirmed_email_repository: Repositorio para emails confirmados
        """
        self._confirmed_email_repository = confirmed_email_repository

    async def execute(self, email_str: str) -> ConfirmEmailResponse:
        """Ejecuta la verificación de confirmación de un email.

        Args:
            email_str: Email del usuario

        Returns:
            Response con el estado de confirmación del email
        """
        email = Email(email_str)
        is_confirmed = await self._confirmed_email_repository.is_email_confirmed(email)

        if is_confirmed:
            return ConfirmEmailResponse(
                email=email.value,
                confirmed=True,
                message="Email confirmado exitosamente",
            )

        return ConfirmEmailResponse(
            email=email.value,
            confirmed=False,
            message="El email aún no ha sido confirmado (OTP no verificado)",
        )

File: backend/goat_listing/src/main/java/com/goat/identity/application/usecases/ConfirmEmailUseCase.java
package com.goat.identity.application.usecases;

import com.goat.identity.domain.entities.User;
import com.goat.identity.domain.valueobjects.Email;
import com.goat.identity.ports.OtpVerificationPort;
import com.goat.identity.ports.UserRepository;

/**
 * Caso de uso para confirmar el email de un usuario.
 * Verifica con el servicio Python si el OTP fue verificado y actualiza el estado.
 */
public class ConfirmEmailUseCase {
    private final UserRepository userRepository;
    private final OtpVerificationPort otpVerificationPort;

    public ConfirmEmailUseCase(UserRepository userRepository, OtpVerificationPort otpVerificationPort) {
        this.userRepository = userRepository;
        this.otpVerificationPort = otpVerificationPort;
    }

    /**
     * Verifica con el servicio Python si el email está confirmado y actualiza el estado en la BD.
     *
     * @param email Email del usuario a verificar
     * @return true si el email fue confirmado y actualizado, false en caso contrario
     */
    public boolean execute(Email email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new IllegalArgumentException("Usuario no encontrado"));

        // Consultar al servicio Python si el OTP fue verificado
        boolean isConfirmed = otpVerificationPort.isEmailConfirmed(email);

        // Si el servicio Python confirma que el OTP fue verificado,
        // actualizar el estado en la base de datos PostgreSQL
        if (isConfirmed) {
            if (!user.getEmailConfirmed()) {
                // Actualizar el estado del usuario en la base de datos
                user.setEmailConfirmed(true);
                userRepository.save(user);
            }
            return true;
        }

        return false;
    }
}

